p2p: Request NOTFOUND transactions immediately from other outbound peers, when possible

sdaftuar opened the pull request:
Currently when a peer asks for a transaction that we cannot (or will not) deliver, we send a NOTFOUND message for the given txid.  However, our software currently ignores the NOTFOUND messages we receive from others -- if other peers have announced a transaction, we wait until the transaction request times out before requesting from our next peer.

Improve this by immediately requesting the transaction from one of the outbound peers who have recently announced it (if any).

sdaftuar committed:
[p2p] Maintain a set of outbound peers

sdaftuar committed:
Request NOTFOUND transactions immediately

Thanks to James O'Beirne for some of the code in this patch.

sdaftuar committed:
Refactor transaction getdata request logic

jamesob commented on src/net_processing.cpp:
```diff
@@ -821,7 +836,21 @@ void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTim
+            g_outbound_peers[peer_loc] = g_outbound_peers.back();
+        }
+        g_outbound_peers.pop_back();
+    }
```
Seems like this could be a lot more succinct with [`vector::erase`](https://en.cppreference.com/w/cpp/container/vector/erase) - is there some reason to do it this way?

sdaftuar replied:
I guess I have an aversion to vector::erase because it requires moving all the other entries.  Possibly that concern doesn't make sense for such a small vector?

But the bigger issue here is why this is a vector at all -- originally I had this as a set, but then I changed it to be a vector in order to make the logic around when to drain the notfound queue for each peer make sense.  In short, I wanted to uniformly at random assign a notfound transaction to an outbound peer that announced it, but I didn't have a good way to do that, so instead I assign it to all peers, but then have a (hacky) counter in SendMessages and each outbound peer will check that counter to decide when to drain its notfound queue.

(I'm hoping that reviewers will give me a better idea of how to implement that logic than what I've got so far...)

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -338,6 +352,11 @@ struct CNodeState {
      */
     struct TxDownloadState {
+        /* Notifications of NOTFOUND transactions (from other peers) which
+         * this peer has recently announced. txid->time in micros
```
This comment seems outdated

sdaftuar replied:
Fixed.

practicalswift commented on src/net_processing.cpp:
```diff
@@ -3081,8 +3156,46 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+                    if (!AlreadyHave(inv)) {
+                        for (auto it = g_outbound_peers.begin(); it != g_outbound_peers.end(); ++it) {
+                            if (*it == pfrom->GetId()) continue;
+                            CNodeState *state = State(*it);
```
This `state` shadows the `state` in the outer scope.

sdaftuar replied:
Fixed.

practicalswift commented on src/net_processing.cpp:
```diff
@@ -724,6 +744,41 @@ void RequestTx(CNodeState* state, const uint256& txid, int64_t nNow) EXCLUSIVE_L
 }
 
+// Given a txid and a peer that we'd like to consider downloading a transaction
+// from: add an appropriate INV nessage to get_data if it is time to request
```
I've got an important nessage for you: this should be "message".

sdaftuar replied:
Fixed.

sdaftuar commented:
Thanks for the nits; I'll address them later, but in the meantime I'm open to suggestions on how to better implement the random assignment of notfound transactions to the outbound peers that have announced them (I think the logic I have here should work, but it's a bit gross).

jamesob commented on src/net_processing.cpp:
```diff
@@ -3879,39 +3992,35 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
+            // every outbound peer gets to drain its notfound queue when the
+            // counter matches its location.
+            assert(g_outbound_peers.size() > 0);
+            if (g_outbound_peers[not_found_counter % g_outbound_peers.size()] == pto->GetId()) {
```
Correct me if I'm wrong here, but isn't this equivalent to doing something like
```cpp
if (GetRandInt(/* nMax= */g_outbound_peers.size()) == 0) {
...
}
```
?

If you think the potential performance difference (which IMO is probably negligible when shoulder-to-shoulder with network IO) is okay, just doing something like that would allow you to keep `g_outbound_peers` a set and preserve the algorithm.

sdaftuar replied:
Doing a 1 in N draw each time would introduce a bias to peers that are earlier in the loop, versus ones that are later in the loop.  Imagine if N=2, and we have two peers, peer0 and peer1.  peer0 would be selected with probability 1/2 + 1/8 + 1/16 + ... = 2/3.  If N=8, I think the first peer would have a 19% chance of being selected.

sdaftuar replied:
I just realized that my hacky implementation here is broken too -- the `not_found_counter` will basically increment by the number of peers we have, so if the number of peers we have is equal the number of outbound peers we have (a common state for non-listening nodes) then we'll always be requesting notfound's from the first peer.

jamesob replied:
I'm not sure I see how there's any bias based on the ordering of peers - the `GetRandInt` thing is an independent trial across peers, though of course we may end up draining multiple peers' `m_not_found` per single [`ThreadMessageHandler` loop](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L1954-L1978) if two peers both get `1/num_outbounds` lucky. Is there a reason we'd want to avoid that?

It seems like there might be a bias based upon [which peer we select to queue the `m_not_found` message on](https://github.com/bitcoin/bitcoin/pull/15505/files#diff-eff7adeaec73a769788bb78858815c91R3184) (peers earlier in `g_outbound_peers` will get selected first more often), but I'm not sure how the order in which we *flush* `m_not_found` makes a difference.

sdaftuar replied:
Ok I redid this to be (hopefully) better, thanks.

jamesob reviewed (commented):
Feel free to ignore the nits, but curious what you think about the use of `GetRandInt` in lieu of the vector hackery.

sdaftuar committed:
[net] Add a global sequence number to SendMessages()

Keep an incrementing counter between loop iterations in ThreadMessageHandler.

This allows for synchronization between loop iterations in SendMessages.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -724,6 +744,41 @@ void RequestTx(CNodeState* state, const uint256& txid, int64_t nNow) EXCLUSIVE_L
+// from: add an appropriate INV nessage to get_data if it is time to request
+// the transaction from the peer, and update the appropriate tx download state
+// (both the global state and the peer's state). Send a getdata message if the
+// get_data vector grows too large.
```
Done.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -3081,8 +3156,46 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+        vRecv >> vInv;
+        if (vInv.size() <= MAX_INV_SZ) {
+            for (CInv &inv : vInv) {
+                if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX) {
```
Will leave as-is -- in this case I like being able to tell that the code block exactly matches the if condition, rather than matching the opposite of an if condition above it that would have caused the loop iteration to be skipped.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -3081,8 +3156,46 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+                if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX) {
+                    state->m_tx_download.m_tx_announced.erase(inv.hash);
+                    auto in_flight_it = state->m_tx_download.m_tx_in_flight.find(inv.hash);
+                    if (in_flight_it == state->m_tx_download.m_tx_in_flight.end()) {
```
Oops, I meant to use the iterator in the erase() call, saving an extra map lookup.  Just did that in the latest commit.

sdaftuar commented:
@jamesob Thanks for the helpful suggestions.  I rewrote this with `g_outbound_peers` being a set, and moving the global sequence number needed for peer selection to be in the interface between net and net_processing.  (It seems to me like the idea of having synchronization between loop iterations in ThreadMessageHandler is something that is occasionally useful in net_processing, so I hope the interface change there is acceptable.)

jamesob reviewed (commented):
Looks good! I'll write up a related functional test this weekend unless you have a burning desire to do so.

naumenkogs commented:
I must be misunderstanding something.

Imagine we have 2 peers:
[p1, p2]
Start with seq_num = 1

SendMessages(p1, 1 % 2 = 1): false
SendMessages(p2, 1 % 2 = 1): false

SendMessages(p1, 2 % 2 = 0): true
SendMessages(p2, 2 % 2 = 0): true

So, p1 will be always called before p2 (and much likely to request something from p1)?

jamesob commented:
@naumenkogs it's not the result of the modulus that is evaluated to determine whether to `notfound` flush for a given peer; `sequence_number % num_outbounds` is used as an index into the set of outbounds which then determines the peer whose turn it is to flush.

sdaftuar commented:
Let's say we have 2 peers, p1 and p2 in `g_outbound_peers`, and let's assume the `std::set` comparator puts them in that order when iterating through `g_outbound_peers`.  Let's also say both of them have something in their respective notfound queues.

When the sequence number is 1, then 1 % 2 == 1, so we look at g_outbound_peers.begin() + 1, and we get p2's nodeid.  When we call SendMessage on p1, the nodeid will not match, so its queue will not be drained.  When we call SendMessages on p2, the condition will be true, and we'll drain the notfound queue.   On the next loop iteration, the conditions flip.

(I'm open to simpler logic that accomplishes the same thing!)

naumenkogs commented:
@sdaftuar @jamesob I see, you're right, thanks.

I was not a big fan of extra sequence_numbers, but based on the discussion above and some thinking I don't see a cleaner way to select a peer in a more fair way. This sequence_number is also likely to be useful somewhere else in future.

utACK

jamesob commented:
utACK https://github.com/bitcoin/bitcoin/pull/15505/commits/f5dbb49c8922cb26d05862b7aacf161bde8ffc5a

I was going to write a functional test for this, but I'm not sure there's a sensible way to do that given that it looks like we can't force `P2PDataStore` clients to be outbound connections for nodes under test. https://github.com/bitcoin/bitcoin/issues/14210

If anyone can think of a good manual test plan, I'm happy to carry it out.

sdaftuar commented:
Closing for now.

sdaftuar commented:
Seems like not much interest here, closing for now.

jkczyz commented:
Rather than using a sequence number to determine which peer's notfound queue to drain, would the following work?

1. When a `notfound` message is received, add each transaction to the notfound queue of a random peer selected from those that have announced the transaction.
1. Unconditionally drain each peer's notfound queue.

This would avoid the added complexity of a sequence number. It would also ensure that each transaction is requested from only one peer.

jnewbery commented:
@sdaftuar what do you think about @jkczyz's suggestion for randomly selecting which peer to request the tx from? Conceptually, I think I prefer it to the sequence counter in the net/net_processing interface.

sdaftuar commented:
> Rather than using a sequence number to determine which peer's notfound queue to drain, would the following work?
> 
> 1. When a `notfound` message is received, add each transaction to the notfound queue of a random peer selected from those that have announced the transaction.
> 2. Unconditionally drain each peer's notfound queue.

The reason I tend to avoid an approach like this is because the logic for whether we actually send a message to a peer is very complex; there are many reasons why we might call SendMessages() on a peer and then ultimately not be able to send a getdata to it (maybe we have too many getdata’s in flight to them, or maybe we’re processing block requests for them, or maybe their send buffer is full blocking communication for a while, etc).  Since there’s no good way to tell whether a getdata would be sent if assigned to a given peer, I think it’s more robust to just signal all of them and let each peer individually figure out if they can make the request.

MarcoFalke commented:
@sdaftuar Are you still working on this to get it rebased or would you prefer some to take it over?

sdaftuar commented:
Please consider this up for grabs.

fanquake commented:
Going to close. @naumenkogs you might be interested in following up here?
